src/libopensc/pkcs15-esteid2025.c

```diff
@@ -24,67 +24,107 @@
 
 #include "common/compat_strlcpy.h"
 
 #include "internal.h"
 #include "opensc.h"
 #include "pkcs15.h"
 
+static int
+is_fin_eid(sc_pkcs15_card_t *p15card)
+{
+	return p15card->card->type == SC_CARD_TYPE_FINEID_2022 ||
+	       p15card->card->type == SC_CARD_TYPE_FINEID_2025;
+}
+
 static int
 sc_pkcs15emu_esteid2025_init(sc_pkcs15_card_t *p15card)
 {
 	sc_card_t *card = p15card->card;
-	u8 *buff;
 	int r, i;
-	size_t field_length = 0, buflen = 9, j;
+	size_t field_length = 0;
 	sc_path_t tmppath;
 	static const u8 prkey_id[2] = {0x01, 0x02};
+	static const u8 pin_authid[3] = {1, 2, 3};
 
-	set_string(&p15card->tokeninfo->label, "ID-kaart");
-	set_string(&p15card->tokeninfo->manufacturer_id, "Thales");
+	if (is_fin_eid(p15card)) {
+		u8 buf[SC_MAX_APDU_BUFFER_SIZE];
+		const u8 *tag;
+		size_t buflen = sizeof(buf), taglen;
 
-	buff = malloc(buflen + 1);
-	if (!buff)
-		LOG_FUNC_RETURN(card->ctx, SC_ERROR_OUT_OF_MEMORY);
+		sc_format_path("5032", &tmppath);
+		r = sc_select_file(card, &tmppath, NULL);
+		LOG_TEST_RET(card->ctx, r, "Selecting CIA path failed");
 
-	/* Read document number to be used as serial */
-	sc_format_path("DFDD5007", &tmppath);
-	r = sc_select_file(card, &tmppath, NULL);
-	if (r < 0) {
-		free(buff);
-		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INVALID_CARD);
-	}
-	r = sc_read_binary(card, 0, buff, buflen, 0);
-	if (r < 0) {
-		free(buff);
-		LOG_FUNC_RETURN(card->ctx, SC_ERROR_INVALID_CARD);
-	}
+		r = sc_read_binary(p15card->card, 4, buf, buflen, 0);
+		LOG_TEST_RET(card->ctx, r, "Reading tokeninfo file failed");
+
+		tag = sc_asn1_find_tag(card->ctx, buf, buflen, 0x04, &taglen);
+		if (tag == NULL)
+			LOG_TEST_RET(card->ctx, SC_ERROR_INVALID_CARD, "Finding document number tag failed");
+
+		for (size_t j = 0; j < taglen; j++) {
+			if (!isalnum(tag[j]))
+				LOG_TEST_RET(card->ctx, SC_ERROR_INVALID_CARD, "Invalid character in document number");
+		}
+
+		free(p15card->tokeninfo->serial_number);
+		p15card->tokeninfo->serial_number = malloc(taglen + 1);
+		if (p15card->tokeninfo->serial_number == NULL)
+			LOG_TEST_RET(card->ctx, SC_ERROR_OUT_OF_MEMORY, "Allocating memory failed");
+		p15card->tokeninfo->serial_number = memcpy(p15card->tokeninfo->serial_number, tag, taglen);
+		p15card->tokeninfo->serial_number[taglen] = '\0';
+	} else {
+		u8 *buf;
+		size_t buflen = 9;
 
-	for (j = 0; j < buflen; j++) {
-		if (!isalnum(buff[j])) {
-			free(buff);
-			LOG_FUNC_RETURN(card->ctx, SC_ERROR_INVALID_CARD);
+		/* Read document number to be used as serial */
+		sc_format_path("DFDD5007", &tmppath);
+		r = sc_select_file(card, &tmppath, NULL);
+		LOG_TEST_RET(card->ctx, r, "Selecting document number file failed");
+
+		buf = malloc(buflen + 1);
+		if (!buf)
+			LOG_TEST_RET(card->ctx, SC_ERROR_OUT_OF_MEMORY, "Allocating memory failed");
+		r = sc_read_binary(card, 0, buf, buflen, 0);
+		if (r < 0) {
+			free(buf);
+			LOG_TEST_GOTO_ERR(card->ctx, r, "Reading document number failed");
+		}
+
+		for (size_t j = 0; j < buflen; j++) {
```

Comment:
this has the same assumptions that we read the full `buflen`. The card can return less data and then we will use uninitialized memory in `buf`, which is UB.
(I know this is just moved from the previous version, but I think its worth fixing once we touch it)